### PR TITLE
[SPARK-11017] [SQL] Support ImperativeAggregates in TungstenAggregate

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch extends TungstenAggregate to support ImperativeAggregate functions. The existing TungstenAggregate operator only supported DeclarativeAggregate functions, which are defined in terms of Catalyst expressions and can be evaluated via generated projections. ImperativeAggregate functions, on the other hand, are evaluated by calling their `initialize`, `update`, `merge`, and `eval` methods.

The basic strategy here is similar to how SortBasedAggregate evaluates both types of aggregate functions: use a generated projection to evaluate the expression-based declarative aggregates with dummy placeholder expressions inserted in place of the imperative aggregate function output, then invoke the imperative aggregate functions and target them against the aggregation buffer. The bulk of the diff here consists of code that was copied and adapted from SortBasedAggregate, with some key changes to handle TungstenAggregate's sort fallback path.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 1]
```diff
@@ -546,6 +546,9 @@ case class HyperLogLogPlusPlus(child: Expression, relativeSD: Double = 0.05)
     AttributeReference(s"MS[$i]", LongType)()
   }
 
+  override val inputAggBufferAttributes: Seq[AttributeReference] =
```
Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: To explain what's going on here: it turns out that the refactoring in #8973 to lift this into `InterpretedAggregate` ended up introducing a bug due to lazy val initialization issues. To avoid this, I just pushed these copies to the leaves of our inheritance hierarchy and removed the default implementation from the parent class.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala [hunk 2]
```diff
@@ -321,7 +321,7 @@ abstract class AggregationIterator(
   // Initializing the function used to generate the output row.
   protected val generateOutput: (InternalRow, MutableRow) => InternalRow = {
     val rowToBeEvaluated = new JoinedRow
-    val safeOutputRow = new GenericMutableRow(resultExpressions.length)
+    val safeOutputRow = new SpecificMutableRow(resultExpressions.map(_.dataType))
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala]: These changes in `AggregationIterator.scala` are just some minor optimizations motivated by things that I noticed while working on the Tungsten version. Here, and down below, I noticed that we might be unnecessarily boxing via our use of GenericMutableRow, hence this change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala [hunk 3]
```diff
@@ -367,7 +368,7 @@ abstract class AggregationIterator(
 
         (currentGroupingKey: InternalRow, currentBuffer: MutableRow) => {
           // Generate results for all expression-based aggregate functions.
-          expressionAggEvalProjection.target(aggregateResult)(currentBuffer)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala]: Here, I noticed that `expressionAggEvalProjection` is always targeted at `aggregateResult`, so I decided to do the targeting once outside of the loop.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 4]
```diff
@@ -134,19 +137,73 @@ class TungstenAggregationIterator(
...
-  // IllegalStateException.
-  private[this] val allAggregateFunctions: Array[DeclarativeAggregate] = {
-    if (!allAggregateExpressions.forall(
-        _.aggregateFunction.isInstanceOf[DeclarativeAggregate])) {
-      throw new IllegalStateException(
-        "Only ExpressionAggregateFunctions should be passed in TungstenAggregationIterator.")
+  // Initialize all AggregateFunctions by binding references, if necessary,
+  // and setting inputBufferOffset and mutableBufferOffset.
+  private def initializeAllAggregateFunctions(
+      startingInputBufferOffset: Int): Array[AggregateFunction2] = {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: This is one of the key changes: we need to be able to re-initialize the imperative aggregate functions after we fall back to the sort-based path.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 5]
```diff
@@ -170,10 +232,27 @@ class TungstenAggregationIterator(
...
 
     val genericMutableBuffer = new GenericMutableRow(bufferRowSize)
-    val unsafeProjection =
-      UnsafeProjection.create(bufferSchema.map(_.dataType))
-    val buffer = unsafeProjection.apply(genericMutableBuffer)
-    initialProjection.target(buffer)(EmptyRow)
+    // TODO(josh): figure out whether we have to use
+    val useUnsafeBuffer = bufferSchema.map(_.dataType).forall(UnsafeRow.isMutable)
+
+    val buffer =  /* if (useUnsafeBuffer) */ {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Whoops, I meant to clean this up. One concern that I have is whether it's safe to expose an UnsafeRow aggregation buffer to InterpretedAggregate functions. Even if UnsafeRow supports in-place updates for all of the data types used by the aggregate function's buffer schema, I'm still slightly worried about what might happen if the imperative aggregate code were to call the generic `MutableRow.update()` method. As long as we can guarantee that `update()` won't be called by any valid imperative aggregate function, then I think it should be fine to just always use UnsafeRow here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 6]
```diff
@@ -170,10 +232,27 @@ class TungstenAggregationIterator(
...
+      val unsafeProjection =
+        UnsafeProjection.create(bufferSchema.map(_.dataType))
+      unsafeProjection.apply(genericMutableBuffer)
+//    } else {
+//      genericMutableBuffer
+    }
+    expressionAggInitialProjection.target(buffer)(EmptyRow)
+    // TODO(josh): this can be done more cleanly
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: I think that I thought that this block of code was a bit messy while I was working on it. Will clean up while addressing any code review comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 7]
```diff
@@ -187,72 +266,124 @@ class TungstenAggregationIterator(
     aggregationMode match {
       // Partial-only
       case (Some(Partial), None) =>
-        val updateExpressions = allAggregateFunctions.flatMap(_.updateExpressions)
-        val updateProjection =
+        val updateExpressions = allAggregateFunctions.flatMap {
+          case ae: DeclarativeAggregate => ae.updateExpressions
+          case agg: AggregateFunction2 => Seq.fill(agg.aggBufferAttributes.length)(NoOp)
+        }
+        val imperativeAggregateFunctions: Array[ImperativeAggregate] =
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: While `imperativeAggregateFunctions` can't as easily be defined at the top of the class (since it would need to be reset when switching to spilling), I think that I _could_ lift it out of this `match` statement and put it at the top of this function to cut down on code duplication a bit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 8]
```diff
@@ -288,17 +419,33 @@ class TungstenAggregationIterator(
         val joinedRow = new JoinedRow()
         val evalExpressions = allAggregateFunctions.map {
           case ae: DeclarativeAggregate => ae.evaluateExpression
-          // case agg: AggregateFunction2 => Literal.create(null, agg.dataType)
+          case agg: AggregateFunction2 => NoOp
         }
-        val expressionAggEvalProjection = UnsafeProjection.create(evalExpressions, bufferAttributes)
+        val expressionAggEvalProjection = newMutableProjection(evalExpressions, bufferAttributes)()
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: I think that the issue here was that some aggregate _did_ call `update()` on the UnsafeRow, leading to problems. I'll go ahead and confirm this, though.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 9]
```diff
@@ -511,8 +658,10 @@ class TungstenAggregationIterator(
     }
     aggregationMode = newAggregationMode
 
+    allAggregateFunctions = initializeAllAggregateFunctions(startingInputBufferOffset = 0)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Here's the other TungstenAggregate-specific change.

**Comment by rxin** [comment]
YAY

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 5 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: When we do hash-based aggregation, the buffer is an UnsafeRow anyway. So, I guess it is fine to let InterpretedAggregate functions to work with UnsafeRows directly. When a developer adds an new InterpretedAggregate function whose aggregate buffer data types are compatible with TungstenAggregate, if he/she uses `update`, he/she will get the exception while testing the function, right? If so, it is safe to say that the code committed in master will not use `update`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala (hunk 1 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: So, if anyone tries to move it back to `ImperativeAggregate`, he/she will see an test failure? Should we add a comment to `ImperativeAggregate` to explain why a leaf class needs to implement this?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala [hunk 10]
```diff
@@ -70,7 +70,6 @@ object Utils {
     // Check if we can use TungstenAggregate.
     val usesTungstenAggregate =
       child.sqlContext.conf.unsafeEnabled &&
-      aggregateExpressions.forall(_.aggregateFunction.isInstanceOf[DeclarativeAggregate]) &&
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala]: Can we also remove it from `planAggregateWithOneDistinct`?

For `planAggregateWithoutPartial`, we can add the flag `usesTungstenAggregate` and whenever possible use `TungstenAggregate` (we can do it in a separate pr.).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 9 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Maybe add a comment at here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 11]
```diff
@@ -134,19 +137,73 @@ class TungstenAggregationIterator(
...
+        case _ =>
+      }
+      mutableBufferOffset += funcWithBoundReferences.aggBufferSchema.length
+      functions(i) = funcWithBoundReferences
+      i += 1
     }
+    functions
+  }
+
+  private[this] var allAggregateFunctions: Array[AggregateFunction2] =
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Since it will be a `var`, let's double check places where we use it and make sure they will be updated after we change it. I did a quick search, seems `allImperativeAggregateFunctionPositions` and `expressionAggInitialProjection` are the only two `val`s that will not be updated after we change `allAggregateFunctions`. But, I think we are fine because these two will not be affected by the change of `startingInputBufferOffset`.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Maybe we should add comments to `allImperativeAggregateFunctionPositions` and `expressionAggInitialProjection` to explain why we do not need to update them after we update `allAggregateFunctions`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 12]
```diff
@@ -288,17 +419,33 @@ class TungstenAggregationIterator(
...
+            .map(_.asInstanceOf[ImperativeAggregate])
+
         (currentGroupingKey: UnsafeRow, currentBuffer: UnsafeRow) => {
           // Generate results for all expression-based aggregate functions.
-          val aggregateResult = expressionAggEvalProjection.apply(currentBuffer)
+          expressionAggEvalProjection(currentBuffer)
+          // Generate results for all imperative aggregate functions.
+          var i = 0
+          while (i < allImperativeAggregateFunctions.length) {
+            aggregateResult.update(
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: This `update` prevent us from using `UnsafeRow`, right?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Oh, right; this was on my list of things to discuss. I think that we can handle this by calling the specific setter methods because we happen to know the data type. We might want to do this by building up an array of functions to perform the update, similar to how we used to do things in CatalystTypeConverters.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Will https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/udaf.scala#L99-L202 help?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 5 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Ah, good point. I'll update to reflect this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala (hunk 10 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala]: Whoops, good catch. I removed this from `planAggregateWithOneDistinct` and it led to a failure in `single distinct column set`:

```
[info]   == Results ==
[info]   !== Correct Answer - 4 ==                                                        == Spark Answer - 4 ==
[info]   ![100.0,0.3333333333333333,1.0,2,99.33333333333333,10.0,2.0]                     [100.0,0.3333333333333333,1.0,2,101.0,10.0,2.0]
[info]   ![110.0,10.0,20.0,null,109.0,11.0,30.0]                                          [110.0,10.0,20.0,null,120.0,11.0,30.0]
[info]   ![120.0,23.333333333333332,-3.3333333333333335,1,122.33333333333333,12.0,20.0]   [120.0,23.333333333333332,-3.3333333333333335,1,96.66666666666667,12.0,20.0]
[info]   ![null,null,3.0,3,null,null,null]                                                [null,null,3.0,3,103.0,null,null] (AggregationQuerySuite.scala:690)
```

Based on the failure mode here, I suspect that another one of the buffer offsets needs to be reset in response to the sort fallback.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala]: This also reminds me: are there planner tests that should be updated to assert that we're actually using the new Tungsten path? Let me check and see.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 12 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: As discussed offline, let's not make this change as part of this patch. There's a good chance that calling UnsafeRow's specific setters via this loop may wind up being just as slow as using a SpecificMutableRow. Note that this operator will still _output_ UnsafeRows, since `resultProjection` outputs them.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala [hunk 13]
```diff
@@ -154,143 +146,150 @@ object Utils {
...
-    // It is safe to call head at here since functionsWithDistinct has at least one
-    // AggregateExpression2.
-    val distinctColumnExpressions =
-      functionsWithDistinct.head.aggregateFunction.children
-    val namedDistinctColumnExpressions = distinctColumnExpressions.map {
-      case ne: NamedExpression => ne -> ne
-      case other =>
-        val withAlias = Alias(other, other.toString)()
-        other -> withAlias
+    // functionsWithDistinct is guaranteed to be non-empty. Even though it may contain more than one
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala]: I did a bit of refactoring here to try to add some assertions to clarify what's going on (namely, show that there is only one column expression for the distinct aggregates).

**Comment by JoshRosen** [comment]
Okay, I have a theory as to what might be causing the test failure. The assertion failure is only occurring in the final aggregation, `finalAndCompleteAggregate`.

Consider the input of `finalAndCompleteAggregate`: this is the output of `partialMergeAggregate`, whose result expressions are

```scala
        groupingAttributes ++
          Seq(distinctColumnAttribute) ++
          partialMergeAggregateExpressions.flatMap(_.aggregateFunction.inputAggBufferAttributes)
```

Here, the input to the `finalAndCompleteAggregate` is a partial aggregate because its attributes are `inputAggBufferAttributes`. If we fall back to sort while evaluating `finalAndCompleteAggregate` then we need to extract the partial aggregate results from the input row rather than calling `processRow` to initialize empty aggregation buffers. I think that this is the first problem: it looks like we're mistakenly calling the `needsProcess` branch instead. In this branch, the `inputBufferOffset`s of the imperative aggregate functions will not have been set properly, which I think is leading to wrong answers.

I tried redefining `needsProcess` as 

```scala
    val needsProcess = (initialInputBufferOffset == 0)
```

but this led to attribute binding errors. Therefore, I think that we might be assigning the wrong attributes to the result of `partialMergeAggregate`. I'm going to try to fix this by changing the attributes of `partialMergeAggregate` and refactoring the logic when falling back to sort.

**Comment by yhuai** [comment]
ah, `assert(initialInputBufferOffset > 0)` failed because when there is not grouping expression and the mode is not partial, initialInputBufferOffset will be 0.

**Comment by yhuai** [comment]
For that failed test, I think the problem is that those two agg operators in the reducer are holding the same reference of those aggregate functions. Essentially, we want to have a way to clone a agg function without changing `aggBufferAttributes` and `inputAggBufferAttributes`. How about we address this issue in a separate PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 14]
```diff
@@ -134,19 +137,74 @@ class TungstenAggregationIterator(
...
+  // Initialize all AggregateFunctions by binding references, if necessary,
+  // and setting inputBufferOffset and mutableBufferOffset.
+  private def initializeAllAggregateFunctions(
+      startingInputBufferOffset: Int): Array[AggregateFunction2] = {
+    var mutableBufferOffset = 0
+    var inputBufferOffset: Int = startingInputBufferOffset
+    val functions = new Array[AggregateFunction2](allAggregateExpressions.length)
+    var i = 0
+    while (i < allAggregateExpressions.length) {
+      val func = allAggregateExpressions(i).aggregateFunction
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: I tried
```
val originalFunc = allAggregateExpressions(i).aggregateFunction
val baos = new ByteArrayOutputStream()
val funcOut = new ObjectOutputStream(baos)
      funcOut.reset()
funcOut.writeObject(originalFunc)
val byteArray = baos.toByteArray
funcOut.close()
val funcIn = new ObjectInputStream((new ByteArrayInputStream(byteArray)))
val func = funcIn.readObject().asInstanceOf[AggregateFunction2]
```
to confirm my guess (multiple operators are holding the same ImperativeAggregate instance). After I add this, all tests of `TungstenAggregationQueryWithControlledFallbackSuite` pass.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Good catch! I wonder whether we can address this treating ImperativeAggregtate instances as immutable and changing the `with*Offset` methods to return new instances. This seems a bit easier / safer than having to remember to clone.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: I guess that this would also help to avoid potential issues caused by executor-side mutable state in the aggregate functions, should they wind up being used multiple times in different parts of a pipeline.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Yeah, it will be good to return a new instance when we call `with*Offset`. For UDAF, I think it may make sense to add a `newInstance` interface so, if a user-defined one has internal states, we can create copies of it  at executor side. There is one difficulty that we need to preserve those aggBufferAttributes and inputAggBufferAttributes' exprIds when we create new instances. So, binding references can work correctly without requiring too many changes. Is there an easy way to do that?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: One thought: instead of trying to preserve  `aggBufferAttributes` and `inputAggBufferAttributes` when returning a new instance, what if we also updated the other expressions which referenced those attributes? Maybe that would be easier.

**Comment by JoshRosen** [comment]
@yhuai, I think that we need to understand why my latest fix works, since I think that the test failure might impact non-distinct aggregates as well (in which case we can't easily split this ticket into the distinct and non-distinct cases).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala [hunk 15]
```diff
@@ -154,143 +146,150 @@ object Utils {
     val aggregateExpressions = functionsWithDistinct ++ functionsWithoutDistinct
     val usesTungstenAggregate =
       child.sqlContext.conf.unsafeEnabled &&
-        aggregateExpressions.forall(
-          _.aggregateFunction.isInstanceOf[DeclarativeAggregate]) &&
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala]: Should we create a PR to include changes we made before we deleted this check? So, we can merge those changes earlier first and then spend time to figure out the current issue. What do you think?

**Comment by yhuai** [comment]
What is the failed test (and the stack trace)?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala (hunk 15 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala]: Perhaps. Let me see if I'm convinced that the other tests are still valid even if we skip this distinct case.

**Comment by JoshRosen** [comment]
To elaborate on the test failures:

All tests pass as of 8695e3f, when the Tungsten code path is not used for imperative distinct aggregates.

As of a09c51bc6936cdd3bbbdd4a7b73735079f62896d, which enables this for distinct aggregates, the "single distinct column set" query fails by returning the wrong answer. Here, I think the problem was that the attribute bindings were being mixed up, causing us to return the wrong columns in one of the projections.

After trying to fix things by copying when changing the buffer offsets, I wound up getting binding errors. My current understanding of attribute binding is that we shouldn't be creating new attributes on executors because this could lead to expression ids being re-used in an incorrect fashion. Given this, I think the cleanest thing to do will be to do the ImperativeAggregate mutable-part/immutable-part interface refactoring that I proposed earlier.

**Comment by JoshRosen** [comment]
I'm fairly concerned about the implications of having mutable state inside of ImperativeAggregates which isn't reset when switching to sort / could be shared across stages of a pipeline, so I'm slightly worried about the idea of reverting back to 8695e3f and merging this PR as of then: it seems like the same underlying problems could exist there, too.

**Comment by JoshRosen** [comment]
See #9093 for an exploration of some sketchy-looking `newExprId` calls which might be involved in the problems that we're experiencing here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 16]
```diff
@@ -481,10 +610,10 @@ class TungstenAggregationIterator(
       // When needsProcess is false, the format of input rows is groupingKey + aggregation buffer.
       // We need to project the aggregation buffer part from an input row.
       val buffer = createNewAggregationBuffer()
-      // The originalInputAttributes are using cloneBufferAttributes. So, we need to use
-      // allAggregateFunctions.flatMap(_.cloneBufferAttributes).
+      // The originalInputAttributes are using inputAggBufferAttributes. So, we need to use
+      // allAggregateFunctions.flatMap(_.inputAggBufferAttributes).
       val bufferExtractor = newMutableProjection(
-        allAggregateFunctions.flatMap(_.inputAggBufferAttributes),
+        originalInputAttributes.drop(initialInputBufferOffset),
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Had a discussion with @JoshRosen offline. Here is the explaining for this change.

Our `ImperativeAggregate` and `DeclarativeAggregate` are quite different. For an `ImperativeAggregate`, its correctness is based on the correctness of `mutableAggBufferOffset` and `inputAggBufferOffset`, and when there is no fallback, it does not rely on attribute ids for binding. To make `mutableAggBufferOffset` and `inputAggBufferOffset` immutable, whenever we call `withNewMutableAggBufferOffset` and `withNewInputAggBufferOffset`, we create a new copy of the `ImperativeAggregate`. Although a new copy contains a new set of attribute ids for `aggBufferAttributes` and `inputAggBufferAttributes`, in most cases, those new ids do not introduce any problem because an `ImperativeAggregate` does not rely on these ids. However, once we fallback to sort-based aggregation and we need to extract input aggregation buffers, we need to have matching attribute ids to make it work. Since  `allAggregateFunctions.flatMap(_.inputAggBufferAttributes)` will have attributes with a new set of ids, we will see a binding error at here. `originalInputAttributes.drop(initialInputBufferOffset)` is fine at here.

**Comment by yhuai** [comment]
I tried the UDAF in https://databricks.com/blog/2015/09/16/spark-1-5-dataframe-api-highlights-datetimestring-handling-time-intervals-and-udafs.html. I tried the following test in `build/sbt -Phive sparkShell`.
```
// Generating data
sc
  .parallelize(1 to 1000, 1)
  .flatMap(i => (1 to 20000).map(j => (i, j, j, j, j, j, j, j, j, j, j)))
  .toDF("i", "j1", "j2", "j3", "j4", "j5", "j6", "j7", "j8", "j9", "j10")
  .write
  .format("parquet")
  .saveAsTable("testAgg")

// Run query.
import org.apache.spark.sql.functions._

val gm = new GeometricMean
sqlContext.udf.register("gm", gm)

sqlContext.sql("set spark.sql.shuffle.partitions=1")
sqlContext.read.load("/user/hive/warehouse/testagg").registerTempTable("testAgg")

val start = System.currentTimeMillis
(1 to 10).foreach { i=>
sqlContext.sql("""
  select
    i,
    gm(j1),
    gm(j2),
    gm(j3),
    gm(j4),
    gm(j5),
    sum(j6),
    sum(j7),
    sum(j8),
    sum(j9),
    sum(j10)
  from testAgg
  group by i""").queryExecution.executedPlan.execute().foreach(x => Unit)
}
println("took " + ((System.currentTimeMillis - start).toDouble / 1000/10))
```
Without this PR, the avg execution time was 31.6s. With this PR, the avg execution time was 27.8s.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 16 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: I'm folding this into the comments right now. Thanks for summarizing.

**Comment by yhuai** [comment]
test this please

**Comment by yhuai** [comment]
LGTM pending jenkins.

**Comment by yhuai** [comment]
I took a look at query plans of dfs used in those `TungstenAggregation*` tests. Those plans look good.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala [hunk 17]
```diff
@@ -154,143 +146,150 @@ object Utils {
...
-      groupingExpressions ++ namedDistinctColumnExpressions.map(_._2)
-    val partialAggregateResult =
+    // 1. Create an Aggregate Operator for partial aggregations.
+    val partialAggregate: SparkPlan = {
+      val partialAggregateExpressions = functionsWithoutDistinct.map(_.copy(mode = Partial))
+      val partialAggregateAttributes =
+        partialAggregateExpressions.flatMap(_.aggregateFunction.aggBufferAttributes)
+      // We will group by the original grouping expression, plus an additional expression for the
+      // DISTINCT column. For example, for AVG(DISTINCT value) GROUP BY key, the the grouping
+      // expressions will be [value, key].
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala]: seems it should be `[key, value]`. Will fix it while merging it.

**Comment by yhuai** [comment]
Thank you @JoshRosen ! I am merging it to master.
